packages/build/src/core/constants.js

```diff
@@ -49,10 +41,73 @@ const getConstants = function ({
     // The Netlify API host
     NETLIFY_API_HOST: apiHost,
   }
-  const constantsA = mapObj(constants, (key, path) => [key, normalizePath(path, buildDir, key)])
+  const constantsA = await addMutableConstants({ constants, buildDir, netlifyConfig })
   return constantsA
 }
 
+// Retrieve constants which might change during the build if a plugin modifies
+// `netlifyConfig` or creates some default directories.
+// Unlike readonly constants, this is called again before each command.
+const addMutableConstants = async function ({
+  constants,
+  buildDir,
+  netlifyConfig: {
+    build: { publish = buildDir, edge_handlers: edgeHandlers },
+    functionsDirectory,
+  },
+}) {
+  const constantsA = {
+    ...constants,
+    // Directory that contains the deploy-ready HTML files and assets generated by the build
+    PUBLISH_DIR: publish,
+    // The directory where function source code lives
+    FUNCTIONS_SRC: functionsDirectory,
+    // The directory where edge handlers source code lives
+    EDGE_HANDLERS_SRC: edgeHandlers,
+  }
+  const constantsB = await addDefaultConstants(constantsA, buildDir)
+  const constantsC = normalizeConstantsPaths(constantsB, buildDir)
+  return constantsC
+}
+
+// Some `constants` have a default value when a specific file exists.
+// Those default values are assigned by `@netlify/config`. However, the build
+// command or plugins might create those specific files, in which case, the
+// related `constant` should be updated, unless the user has explicitly
+// configured it.
+const addDefaultConstants = async function (constants, buildDir) {
+  const newConstants = await Promise.all(
+    DEFAULT_PATHS.map(({ constantName, defaultPath }) =>
+      addDefaultConstant({ constants, constantName, defaultPath, buildDir }),
+    ),
+  )
+  return Object.assign({}, constants, ...newConstants)
```

Comment:
Is there a reason to use `Object.assign` here instead of object spread?

Reply:
`newConstants` is an array of objects, not a single object, so `{ ...constants, ...newConstants }` would not work.

Reply:
Ah, of course. Thanks!